### PR TITLE
Fixed framework name attribute in mapping file.

### PR DESCRIPTION
**Pull request by popovaan**
Root cause analysis: 
For some frameworks original operation name was saved incorrectly. Before adding to "fw_tensor_debug_info" it was passed to graph.unique_id() which added id in case if graph has a node with the same name. 
Another issue was that in mxnet models tensor names of output nodes were not saved to IR. This happed due to absence of output edges on output nodes where tensor information is stored. 
Solution: 
Store original framework name in "fw_tensor_debug_info" attribute without passing through graph.unique_id().
Add identity nodes to output nodes in mxnet loader for storing tensor information. These nodes are removed when Result nodes are added to graph.

Ticket: 52131

Code:
* [x]  Comments
* [x]  Code style (PEP8)
* [x]  Transformation generates reshape-able IR
* [x]  Transformation preserves original framework node names

Validation:
* [x]  Unit tests
* [x]  Framework operation tests - N/A
* [x]  Transformation tests - N/A
* [x]  Model Optimizer IR Reader check

Documentation:
* [x]  Supported frameworks operations list - N/A
* [x]  Guide on how to convert the **public** model - N/A
* [x]  User guide update - N/A

### CODE REVIEW

File: model-optimizer/mo/front/extractor.py [hunk 1]
```diff
@@ -753,7 +753,7 @@ def add_outputs_identity(graph: Graph, outputs: list, add_edge: callable, params
     for output in outputs:
         fake_node_name = graph.unique_id(output)
         graph.add_node(fake_node_name, name=fake_node_name, identity=True, kind='op', op='Identity',
-                       infer=None, needs_removal=True)
+                       infer=None, needs_removal=True, symbol_dict={'op': 'Identity'})
```
Comment (mvafin) [inline model-optimizer/mo/front/extractor.py:756]: nit: There is an operation in MO made for such a purpose, called FakeOutput, I would suggest using it for consistency.

**Review by mvafin — COMMENTED** [review]
Generally looks good to me

File: model-optimizer/mo/front/extractor.py (hunk 1 above)
Comment (popovaan) [reply model-optimizer/mo/front/extractor.py:756]: FakeOutput cannot be used here because there is a transformation in MO that renames output nodes connected with FakeOutput nodes and sets output name to be tensor name.  And it is not what needed here, we need to keep original output names.

File: model-optimizer/mo/front/mxnet/extractors/utils.py [hunk 2]
```diff
@@ -101,21 +101,25 @@ def has(self, key):
...
+        used_indices.append(src_node_id[0])
+    return edge_list, used_indices
+
+
+def create_mxnet_edge(src_node_id: str, dst_node_id: str, src_port: int, dst_port: int, framework_name: str):
+    edge_attrs = {
+        'in': src_port,
+        'out': dst_port,
+        # debug anchor for framework name, out port and tensor name
+        'fw_tensor_debug_info': [(framework_name, dst_port, framework_name)],
```
Comment (sadolini) [inline model-optimizer/mo/front/mxnet/extractors/utils.py:117]: In caffe 2 names used: bottom and saved layer_name. Here only one. Why? Also add parameter just to return it looks strange for me.

File: model-optimizer/mo/front/mxnet/loader.py [hunk 3]
```diff
@@ -106,12 +109,25 @@ def symbol2nx(graph, model_nodes, model_params, input_names: str = ''):
         graph.add_node(node_name, **symbol_attrs(node))
         graph.node[node_name].update(common_mxnet_fields(Node(graph, node_name)))
         index_node_keys[i] = node_name
+        fw_name_map[node_name] = node['name']
```
Comment (sadolini) [inline model-optimizer/mo/front/mxnet/loader.py:112]: Does it make sense to use node.soft_get('name', node.id)?

File: model-optimizer/mo/front/mxnet/extractors/utils.py (hunk 2 above)
Comment (popovaan) [reply model-optimizer/mo/front/mxnet/extractors/utils.py:117]: 'fw_tensor_debug_info' should contain (framework operation name, destination port, tensor name). In caffe tensor names are stored in "bottom" attribute and there is also an operation name. But in mxnet there are no tensor names, only operation names. So the operation name is used twice.

File: model-optimizer/mo/front/mxnet/loader.py (hunk 3 above)
Comment (popovaan) [reply model-optimizer/mo/front/mxnet/loader.py:112]: Yes, I think it makes sense. But I think the the default value should be empty string, as node.id will be randomly generated id that does not correspond to the framework name.

Comment (popovaan) [reply model-optimizer/mo/front/mxnet/loader.py:112]: We can't use soft_get() here as node is a dictionary obtained from loading .json model file.

**Review by sadolini — APPROVED** [review]

**Review by mvafin — APPROVED** [review]

File: model-optimizer/mo/front/caffe/loader.py [hunk 4]
```diff
@@ -280,6 +280,7 @@ def caffe_pb_to_nx(graph, proto, model):
                 input_dims.append(np.array(list(dims), dtype=np.int64))
                 input_names.append(layer.name)
 
+        layer_name = layer.name
         layer.name = graph.unique_id(layer.name)
```
Comment (lazarevevgeny) [inline model-optimizer/mo/front/caffe/loader.py:284]: We should not modify the `layer` object. Instead I recommend to remove `layer_name` variable and use the following approach:

```
node_id = graph.unique_id(layer.name)
graph.add_node(node_id, pb=layer, model_pb=model_layer, kind='op', type='Parameter')
```

and update the places where `layer.name` is used in context of our graph node (line 289 and 296) to use `node_id` variable.

Comment (lazarevevgeny) [reply model-optimizer/mo/front/caffe/loader.py:284]: Also in line 310 we use `input_names` list containing original `layer.name` values. But we probably should use the corresponding `node_id` there. Please, double check.

File: model-optimizer/mo/front/mxnet/extractors/utils.py [hunk 5]
```diff
@@ -101,21 +101,25 @@ def has(self, key):
...
-            'in': in_port,
-            'out': dest_port,
-            # debug anchor for framework name, out port and tensor name
-            'fw_tensor_debug_info': [(index_node_key[src_node], src_node_id[1], nodes_list[src_node]['name'])],
-            'in_attrs': ['in'],
-            'out_attrs': ['out'],
-            'data_attrs': ['fw_tensor_debug_info']
-        }
-        edge = (index_node_key[src_node], index_node_key[node_id], edge_attrs)
+        edge = create_mxnet_edge(index_node_key[src_node_id[0]], index_node_key[node_id], in_port, src_node_id[1],  nodes_list[src_node_id[0]]['name'])
```
Comment (lazarevevgeny) [inline model-optimizer/mo/front/mxnet/extractors/utils.py:106]: redundant space character.

File: model-optimizer/mo/front/onnx/loader.py [hunk 6]
```diff
@@ -109,7 +109,7 @@ def protobuf2nx(graph: Graph, pb):
                     'out': src_port,
                     'in': 0,
                     'name': out,
-                    'fw_tensor_debug_info': [(id, src_port, out)],
+                    'fw_tensor_debug_info': [(node_id(node), src_port, out)],
```
Comment (lazarevevgeny) [inline model-optimizer/mo/front/onnx/loader.py:112]: Why this change is needed? The `id` is already the uniquely generated id of the graph node (line 77). And this `id` is actually the right identifier of the node which should be stored in the fw_tensor_debug_info

File: model-optimizer/mo/front/mxnet/loader.py [hunk 7]
```diff
@@ -106,12 +109,25 @@ def symbol2nx(graph, model_nodes, model_params, input_names: str = ''):
...
 
+    used_indices_list = []
```
Comment (lazarevevgeny) [inline model-optimizer/mo/front/mxnet/loader.py:114]: It is more efficient to use `set` instead of `list`.

File: model-optimizer/mo/front/mxnet/loader.py [hunk 8]
```diff
@@ -106,12 +109,25 @@ def symbol2nx(graph, model_nodes, model_params, input_names: str = ''):
...
     for i, attrs in enumerate(model_nodes):
         node = attrs
-        edges = get_mxnet_node_edges(node, i, list(model_nodes), index_node_keys)
+        edges, used_indices = get_mxnet_node_edges(node, i, list(model_nodes), index_node_keys)
         if len(edges) > 0:
             graph.add_edges_from(edges)
+        used_indices_list += used_indices
+
+    output_ids = [index_node_keys[node_id] for node_id in set(range(1, len(model_nodes))) - set(used_indices_list)]
```
Comment (lazarevevgeny) [inline model-optimizer/mo/front/mxnet/loader.py:122]: Why do we skip element with index 0?

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]

**Comment by lazarevevgeny** [comment]
@iimironov , please, review the MxNet part.

**Comment by lazarevevgeny** [comment]
@popovaan, also we need init tests to cover new changes. Consider adding MO tests in the tests-private repo.

File: model-optimizer/mo/front/mxnet/loader.py (hunk 8 above)
Comment (popovaan) [reply model-optimizer/mo/front/mxnet/loader.py]: It is a mistake. Corrected.

File: model-optimizer/mo/front/onnx/loader.py (hunk 6 above)
Comment (popovaan) [reply model-optimizer/mo/front/onnx/loader.py]: As we discussed we need to use node_id(node). Refactored this code for better readability.

File: model-optimizer/mo/front/caffe/loader.py (hunk 4 above)
Comment (popovaan) [reply model-optimizer/mo/front/caffe/loader.py]: Done.

File: model-optimizer/mo/front/mxnet/extractors/utils.py [hunk 9]
```diff
@@ -101,21 +101,26 @@ def has(self, key):
...
-        edge = (index_node_key[src_node], index_node_key[node_id], edge_attrs)
+        edge = create_mxnet_edge(index_node_key[src_node_id[0]], index_node_key[node_id], in_port, src_node_id[1],
+                                 nodes_list[src_node_id[0]]['name'])
         edge_list.append(edge)
-    return edge_list
+        used_indices.add(src_node_id[0])
+    return edge_list, used_indices
+
+
+def create_mxnet_edge(src_node_id: str, dst_node_id: str, src_port: int, dst_port: int, framework_name: str):
```
[the remark this answers is not in the document]
Comment (lazarevevgeny) [reply model-optimizer/mo/front/mxnet/extractors/utils.py:113]: I don't think that is really needed, because the function itself does nothing but return a tuple of parameters. We agreed to add higher level tests which will cover .mapping file generation.

**Review by lazarevevgeny — APPROVED** [review]

**Review by iimironov — APPROVED** [review]
